web/Services/PostcodeApiService.cs

```diff
@@ -0,0 +1,97 @@
+﻿using BricksAndHearts.Database;
+using Microsoft.EntityFrameworkCore;
+using Newtonsoft.Json;
+
+namespace BricksAndHearts.Services;
+
+public interface IPostcodeApiService
+{
+    public Task<List<PropertyDbModel>?> SortPropertiesByLocation(string postalCode);
+}
+
+public class PostcodeApiService : IPostcodeApiService
+{
+    private readonly HttpClient _client = new();
+    private readonly ILogger<PostcodeApiService> _logger;
+    private readonly BricksAndHeartsDbContext _dbContext;
+
+    public PostcodeApiService(ILogger<PostcodeApiService> logger, BricksAndHeartsDbContext dbContext, HttpClient? client = null)
+    {
+        _logger = logger;
+        _dbContext = dbContext;
+        if (client != null)
+        {
+            _client = client;
+        }
+    }
+
+    public async Task<string> MakeApiRequestToPostcodeApi(string postalCode)
+    {
+        var escapedPostalCode = Uri.EscapeDataString(postalCode);
+        var uri =
+            $"https://api.postcodes.io/postcodes/{escapedPostalCode}";
+        var responseBody = string.Empty;
+        _logger.LogInformation("Making API Request to {Uri}", uri);
+        try
+        {
+            var response = await _client.GetAsync(uri);
+            response.EnsureSuccessStatusCode();
+            responseBody = await response.Content.ReadAsStringAsync();
+            _logger.LogInformation("Successful API request to {Uri}", uri);
+        }
+        catch (Exception e)
+        {
+            _logger.LogWarning("Message :{Message} ", e.Message);
+        }
+
+        return responseBody;
+    }
+
+    public PostcodeioResponseModel TurnResponseBodyToModel(string responseBody)
+    {
+        if (string.IsNullOrEmpty(responseBody))
+        {
+            return new PostcodeioResponseModel();
+        }
+
+        var postcodeResponse =
+            JsonConvert.DeserializeObject<PostcodeioResponseModel>(responseBody);
+        return postcodeResponse!;
+    }
+
+    public async Task<List<PropertyDbModel>?> SortPropertiesByLocation(string postalCode)
```

Comment:
Minor, but I'd expect this to be a method in property service, that then calls PostcodeApiService to get the location.

Reply:
So I tried doing this but ended up too much codes is being affected especially the testbase. I guess I could make the methods in PostcodeApiService static and access them in PropertyService directly using PostcodeApiService.[method] rather than constructing an instance of PropertyService? Because PropertyService also requires a logger and I'm not too sure as to how to do it properly without breaking anything...